crates/matrix-sdk-crypto/src/session_manager/group_sessions/mod.rs

```diff
@@ -746,9 +756,173 @@ impl GroupSessionManager {
 
         Ok(requests)
     }
+
+    /// Collect the devices belonging to the given user, and send the details of
+    /// a room key bundle to those devices.
+    ///
+    /// Returns a list of to-device requests which must be sent.
+    #[instrument(skip(self, bundle_data))]
+    pub async fn share_room_key_bundle_data(
+        &self,
+        user_id: &UserId,
+        collect_strategy: &CollectStrategy,
+        bundle_data: RoomKeyBundleContent,
+    ) -> OlmResult<Vec<ToDeviceRequest>> {
+        // Only allow conservative sharing strategies
+        let collect_strategy = match collect_strategy {
+            CollectStrategy::AllDevices | CollectStrategy::ErrorOnVerifiedUserProblem => {
+                warn!("Ignoring request to use unsafe sharing strategy {:?} for room key history sharing", collect_strategy);
+                &CollectStrategy::IdentityBasedStrategy
+            }
+            CollectStrategy::IdentityBasedStrategy | CollectStrategy::OnlyTrustedDevices => {
+                collect_strategy
+            }
+        };
+
+        let mut changes = Changes::default();
+
+        let CollectRecipientsResult { devices, .. } =
+            share_strategy::collect_recipients_for_share_strategy(
+                &self.store,
+                iter::once(user_id),
+                collect_strategy,
+                None,
+            )
+            .await?;
+
+        let devices = devices.into_values().flatten().collect();
+        use crate::types::events::EventType;
+        let event_type = bundle_data.event_type().to_owned();
+        let (requests, _) = self
+            .encrypt_key_content_for_devices(devices, &event_type, bundle_data, &mut changes)
+            .await?;
+
+        // TODO: figure out what to do with withheld devices
+
+        // Persist any changes we might have collected.
+        if !changes.is_empty() {
+            let session_count = changes.sessions.len();
+
+            self.store.save_changes(changes).await?;
+
+            trace!(
+                session_count = session_count,
+                "Stored the changed sessions after encrypting an room key"
+            );
+        }
+
+        Ok(requests)
+    }
+
+    /// Encrypt the given content for the given devices and build to-device
+    /// requests to send the encrypted content to them.
+    ///
+    /// Returns a tuple containing (1) the list of to-device requests, and (2)
+    /// the list of devices that we could not find an olm session for (so
+    /// need a withheld message).
+    async fn encrypt_key_content_for_devices(
```

Comment:
This method can encrypt any content, right? Why is `_key_` part of the name then?

Reply:
it can, though my thinking was that the logic for deciding which devices we would encrypt to, and for generating "withheld" events, would probably only be relevant for "key" data.

Happy to remove `_key` if you think it should go.

Reply:
Hmm, the VOIP team were mentioning that they were interested in a function that can encrypt arbitrary data to multiple devices.

But perhaps they are also sending keys over?

Though if we make this available to the general public, the `key` in there might confuse people.

I think I prefer the removal of `key` slightly but not really enough to make you change things.

Reply:
I need to do some rebasing so I'll drop the `_key` while I'm at it.